src/table/table-role/table-role-helper.ts

```diff
@@ -0,0 +1,98 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+type SortingStatus = 'sortable' | 'ascending' | 'descending';
+
+const stateToAriaSort = {
+  sortable: 'none',
+  ascending: 'ascending',
+  descending: 'descending',
+} as const;
+const getAriaSort = (sortingState: SortingStatus) => stateToAriaSort[sortingState];
+
+/**
+ * Depending on its content the table can have different semantic representation which includes the
+ * ARIA role of the table component ("table", "grid", "treegrid") but also roles and other semantic attributes
+ * of the child elements. The TableRole helper encapsulates table's semantic structure.
+ */
+export class TableRole {
```

Comment:
We use React and this encourages us to use pure functions instead of mutable objects.

Especially, objects that you propagate across components. Does not happen in your code, because all `assign` functions do not modify `this`, but it paves the way for bugs.

Safer option would be to propagate the role directly, without helper

Reply:
On a higher level the it looks like this task is suitable to be solved with contexts

```js
<TableContext role={hasEditableCells ? 'grid' : 'table'} />
```

```js
const {getTableRowProps} = useTableContext();
```

---

We have several more props passing down the same way, so maybe table context may hold more than just this

Reply:
I think contexts is an overkill here - instead I am passing a stable object's reference - it is simpler and no pub/sub is required. What do you think if I make the tableRole property readonly?

Reply:
You could just propagate a string role instead of the object

Reply:
I think of that class as a way of binding the helper functions to the tableRole value - surely can have a module with the functions that would accept the tableRole as an additional argument, but I like the semantics of the helper more. I think it is a good idea to ensure the internal state never changes after initialisation.

Reply:
Classes are not suitable for this job. If you want to know more details, welcome to my DM

Reply:
I updated the helper to be a function instead, please check.